distributed/tests/test_stories.py

```diff
@@ -0,0 +1,170 @@
+import pytest
+
+import dask
+
+from distributed import Worker
+from distributed.comm import CommClosedError
+from distributed.utils_test import assert_story, assert_valid_story, gen_cluster, inc
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_scheduler_story_stimulus_success(c, s, a):
+    f = c.submit(inc, 1)
+    key = f.key
+
+    await f
+
+    stories = s.story(key)
+
+    stimulus_ids = {s[-2] for s in stories}
+    # Two events
+    # - Compute
+    # - Success
+    assert len(stimulus_ids) == 2
+    assert_story(
+        stories,
+        [
+            (key, "released", "waiting", {key: "processing"}),
+            (key, "waiting", "processing", {}),
+            (key, "processing", "memory", {}),
+        ],
+    )
+
+    await c.close()
+
+    stories_after_close = s.story(key)
+    assert len(stories_after_close) > len(stories)
+
+    stimulus_ids = {s[-2] for s in stories_after_close}
+    # One more event
+    # - Forget / Release / Free since client closed
+    assert len(stimulus_ids) == 3
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_scheduler_story_stimulus_retry(c, s, a):
+    def task():
+        assert dask.config.get("foo")
+
+    with dask.config.set(foo=False):
+        f = c.submit(task)
+        with pytest.raises(AssertionError):
+            await f
+
+    with dask.config.set(foo=True):
+        await f.retry()
+        await f
+
+    story = s.story(f.key)
+    stimulus_ids = {s[-2] for s in story}
+    # Four events
+    # - Compute
+    # - Erred
+    # - Compute / Retry
+    # - Success
+    assert len(stimulus_ids) == 4
+
+    assert_story(
+        story,
+        [
+            # Erred transitions via released
+            (f.key, "processing", "erred", {}),
+            (f.key, "erred", "released", {}),
+            (f.key, "released", "waiting", {f.key: "processing"}),
+        ],
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_client_story(c, s, a):
+    f = c.submit(inc, 1)
+    assert await f == 2
+    story = await c.story(f.key)
+
+    # Every event should be prefixed with it's origin
+    # This changes the format compared to default scheduler / worker stories
```

Comment:
I like that if we're doing something that's pretty specific that we explain why we're doing what we're doing.  This helps ...

1. Give me the reviewer confidence that this test was intentional (rather than just copy-pasting current output)
2. Gives futures developers context to debug this test when it fails on them

:+1: